library/core/src/ptr/non_null.rs

```diff
@@ -631,6 +631,12 @@ impl<T: ?Sized> NonNull<T> {
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_allow_const_fn_unstable(unchecked_neg)]
+    #[requires(
+        count.checked_mul(core::mem::size_of::<T>()).is_some() &&
+        count * core::mem::size_of::<T>() <= isize::MAX as usize && 
+        kani::mem::same_allocation(self.as_ptr() as *const(), self.as_ptr().sub(count) as *const())
```

Comment:
```suggestion
        kani::mem::same_allocation(self.as_ptr(), self.as_ptr().wrapping_sub(count))
```

Reply:
Suggesting `wrapping_sub` because `sub` will invoke UB if we cross object boundaries, so if you execute `sub` and it's not in the same allocation, the contract will cause UB. Instead, use `wrapping_sub` so that it produces a pointer, and then `same_allocation` can reason about whether that pointer indeed refers to the same allocation as `self`.

Reply:
resolved